core/src/main/java/org/elasticsearch/action/bulk/BulkItemResponse.java

```diff
@@ -71,12 +83,80 @@ public XContentBuilder toXContent(XContentBuilder builder, Params params) throws
+        ensureExpectedToken(XContentParser.Token.START_OBJECT, token, parser::getTokenLocation);
+
+        token = parser.nextToken();
+        ensureExpectedToken(XContentParser.Token.FIELD_NAME, token, parser::getTokenLocation);
+
+        String currentFieldName = parser.currentName();
+        token = parser.nextToken();
+
+        final OpType opType = OpType.fromString(currentFieldName);
+        ensureExpectedToken(XContentParser.Token.START_OBJECT, token, parser::getTokenLocation);
+
+        DocWriteResponse.ParsingContext context = null;
+        CheckedConsumer<XContentParser, IOException> itemParser = null;
+        Supplier<? extends DocWriteResponse> itemSupplier = null;
+
+        if (opType == OpType.INDEX || opType == OpType.CREATE) {
+            final IndexResponse.ParsingContext parsingContext = new IndexResponse.ParsingContext();
+            context = parsingContext;
+            itemParser = (indexParser) -> IndexResponse.parseXContentFields(indexParser, parsingContext);
+            itemSupplier = () -> IndexResponse.fromParsingContext(parsingContext);
+
+        } else if (opType == OpType.UPDATE) {
+            final UpdateResponse.ParsingContext parsingContext = new UpdateResponse.ParsingContext();
+            context = parsingContext;
+            itemParser = (updateParser) -> UpdateResponse.parseXContentFields(updateParser, parsingContext);
+            itemSupplier = () -> UpdateResponse.fromParsingContext(parsingContext);
+
+        } else if (opType == OpType.DELETE) {
+            final DeleteResponse.ParsingContext parsingContext = new DeleteResponse.ParsingContext();
+            context = parsingContext;
+            itemParser = (deleteParser) -> DeleteResponse.parseXContentFields(deleteParser, parsingContext);
+            itemSupplier = () -> DeleteResponse.fromParsingContext(parsingContext);
+        } else {
+            throwUnknownField(currentFieldName, parser.getTokenLocation());
+        }
+
+        ElasticsearchException exception = null;
+        while ((token = parser.nextToken()) != XContentParser.Token.END_OBJECT) {
+            if (token == XContentParser.Token.FIELD_NAME) {
+                currentFieldName = parser.currentName();
+            }
+
+            if (STATUS.equals(currentFieldName)) {
+                // ignoring status field
+                continue;
+            }
+
+            if (ERROR.equals(currentFieldName)) {
+                if (token == XContentParser.Token.START_OBJECT) {
+                    exception = ElasticsearchException.fromXContent(parser);
+                }
+                continue;
+            }
+
+            itemParser.accept(parser);
+        }
+
+        ensureExpectedToken(XContentParser.Token.END_OBJECT, token, parser::getTokenLocation);
+        token = parser.nextToken();
+        ensureExpectedToken(XContentParser.Token.END_OBJECT, token, parser::getTokenLocation);
+
+        // Id cannot be parsed at this level
+        int id = -1;
```

Comment:
Just out of interest: Could we later set the id on the side where this `fromXContent()` is called? It seems to be the position of the item in the bulk response? Maybe this could be passed in as a second argument to the XContentParser?